src/app/rosetta/lib/account.ml

```diff
@@ -28,13 +29,173 @@ module Get_balance =
+    end
+
+    let query =
+      (* TODO: Need to handle case of multiple commands in same block. (We only care about the latest one.) *)
+      (* TODO: Can we remove some of these fields from the chain select? *)
+      Caqti_request.find_opt
+        Caqti_type.(tup2 string int64)
+        Result.typ
+        {|
+WITH RECURSIVE chain AS (
+  (SELECT id, state_hash, parent_id, height, timestamp
+  FROM blocks b
+  WHERE height = (select MAX(height) from blocks)
+  ORDER BY timestamp ASC
+  LIMIT 1)
+
+  UNION ALL
+
+  SELECT b.id, b.state_hash, b.parent_id, b.height, b.timestamp FROM blocks b
+  INNER JOIN chain
+  ON b.id = chain.parent_id AND chain.id <> chain.parent_id
+),
+
+relevant_internal_commands AS (
+  SELECT
+    ic.id,
+    ic.receiver_id
+  FROM internal_commands ic
+  INNER JOIN public_keys pk ON pk.id = ic.receiver_id
+  WHERE pk.value = $1
+),
+
+relevant_user_commands AS (
+  SELECT
+    uc.id,
+    uc.fee_payer_id,
+    uc.source_id,
+    uc.receiver_id
+  FROM user_commands uc
+  INNER JOIN public_keys pk1 ON pk1.id = uc.fee_payer_id
+  INNER JOIN public_keys pk2 ON pk2.id = uc.source_id
+  INNER JOIN public_keys pk3 ON pk3.id = uc.receiver_id
+  WHERE
+    pk1.value = $1 OR
+    pk2.value = $1 OR
+    pk3.value = $1
+),
+
+relevant_internal_block_balances AS (
+  SELECT
+    block_internal_command.block_id,
+    block_internal_command.receiver_balance AS balance
+  FROM blocks_internal_commands block_internal_command
+  INNER JOIN relevant_internal_commands internal_command
+  ON block_internal_command.internal_command_id = internal_command.id
+),
+
+relevant_user_block_balances AS (
+  SELECT
+    block_user_command.block_id,
+    CASE WHEN pk1.value = $1 THEN block_user_command.fee_payer_balance
+         WHEN pk2.value = $1 THEN block_user_command.source_balance
+         ELSE block_user_command.receiver_balance
+    END AS balance
+  FROM blocks_user_commands block_user_command
+
+  INNER JOIN relevant_user_commands user_command ON block_user_command.user_command_id = user_command.id
+
+  INNER JOIN public_keys pk1 ON pk1.id = user_command.fee_payer_id
+  INNER JOIN public_keys pk2 ON pk2.id = user_command.source_id
+  INNER JOIN public_keys pk3 ON pk3.id = user_command.receiver_id
+),
+
+relevant_block_balances AS (
+  (SELECT block_id, balance from relevant_internal_block_balances)
+  UNION
+  (SELECT block_id, balance from relevant_user_block_balances)
+)
+
+SELECT
+  chain.height AS block_height,
+  balance
+FROM
+chain
+JOIN relevant_block_balances ribb ON chain.id = ribb.block_id
+WHERE chain.height <= $2
+ORDER BY (chain.height, -timestamp) DESC
+LIMIT 1
+      |}
+
+    let run (module Conn : Caqti_async.CONNECTION) start_block_height address =
+      Conn.find_opt query (address, start_block_height)
+  end
+
+  let run (module Conn : Caqti_async.CONNECTION) block_query address =
+    let module M = struct
+      include Deferred.Result
+    end in
+    let open M.Let_syntax in
```

Comment:
nit: `let open Deferred.Result.Let_syntax in`

Reply:
Thanks, this is a remnant of copy->paste->refactor and I could tell it needed to be better but I wasn't sure how to clean it up

Reply:
If I make your change I need `Deferred.Result.fail` instead of `M.fail` below. Is that right or should I `open Deferred` or `open Deferred.Result` too?